Problem: no compact storage for archive node

yihuang opened the pull request:
Closes: #704 

Solution:
- patch iavl to support extracting state changes
- implement versiondb using rocksdb user-defined timestamp feature
- ~~now librocksdb is always required to build cronosd~~
  versiondb is only enabled if binary is built with rocksdb backend.
- implement cli commands to manage the migration process:

```
$ ./build/cronosd changeset
dump and manage change sets files and ingest into versiondb

Usage:
  cronosd changeset [command]

Available Commands:
  convert-to-sst Convert change set files to versiondb/rocksdb sst files, which can be ingested into versiondb later
  dump           Extract changesets from iavl versions, and save to plain file format
  ingest-sst     Ingest sst files into versiondb
  print          Pretty-print the content of change set file
  to-versiondb   Feed change set files into versiondb
  verify         Replay the input change set files in order to rebuild iavl tree in memory and output root hash, user can compare the root hash against the on chain hash
```

- [Review Guides](#issuecomment-1403201363)

- [VersionDB Documentation](https://github.com/crypto-org-chain/cronos/blob/45162664a0f7bf25f4a3c14db01baf45d737a98a/versiondb/README.md)

- [The Future Plan for IAVL Tree Itself](https://github.com/crypto-org-chain/cronos/blob/45162664a0f7bf25f4a3c14db01baf45d737a98a/versiondb/memiavl/README.md)

- [Performance Numbers in Testnet Run](#issuecomment-1403505389)

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻


# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

tomtau commented on cmd/cronosd/cmd/changeset.go:
```diff
@@ -0,0 +1,687 @@
+				endVersion = latestVersion + 1
+			}
+
+			tasksChan := make(chan *dumpTask, 2^32)
```
it looks like, this meant to be 2 << 32? 2^32==34

yihuang replied:
ah, yes

yihuang replied:
it's reallocate the buffer, so I change it to 2048, should be big enough.

tomtau reviewed (commented):
will there be any fallback for other db backends?

yihuang commented:
> will there be any fallback for other db backends?

that would be the original proposal, if there are any interests for that.
user can still specify different db backends for tm and iavl db though, just the versiondb will be always rocksdb.

yihuang commented on go.mod:
```diff
@@ -200,10 +204,15 @@ replace (
 	github.com/tendermint/tendermint => github.com/tendermint/tendermint v0.34.24-0.20221110131553-ec471ba27efd
 
+	// https://github.com/yihuang/tm-db/tree/rocksdb-v7
+	github.com/tendermint/tm-db => github.com/yihuang/tm-db v0.0.0-20221223014825-64db9113ab50
```
https://github.com/tendermint/tm-db/pull/312

yihuang commented on go.mod:
```diff
@@ -200,10 +204,15 @@ replace (
 	github.com/gin-gonic/gin => github.com/gin-gonic/gin v1.7.0
 	github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
 
+	github.com/linxGnu/grocksdb => github.com/yihuang/grocksdb v1.7.11-0.20221223014353-a4fa14d1759b
```
https://github.com/linxGnu/grocksdb/pull/100
fix easier build.

yihuang commented on go.mod:
```diff
@@ -200,10 +204,15 @@ replace (
 	github.com/tendermint/tendermint => github.com/tendermint/tendermint v0.34.24-0.20221110131553-ec471ba27efd
 
+	// https://github.com/yihuang/tm-db/tree/rocksdb-v7
+	github.com/tendermint/tm-db => github.com/yihuang/tm-db v0.0.0-20230103092224-3f21822ed03e
```
https://github.com/tendermint/tm-db/pull/312
migrate to grocksdb to support rocksdb v7.

yihuang commented on go.mod:
```diff
@@ -192,6 +194,9 @@ replace (
 	github.com/cosmos/cosmos-sdk => github.com/cosmos/cosmos-sdk v0.46.7
+
+	// https://github.com/yihuang/iavl/tree/dump_changeset
+	github.com/cosmos/iavl => github.com/yihuang/iavl v0.16.1-0.20230103093747-1b1febf19315
```
https://github.com/cosmos/iavl/pull/654
Add API `TraverseStateChanges`.

yihuang commented:
- `versiondb/*.go`, the abstract versiondb implementation, defines the `VersionDB` interface, and implements the cosmos-sdk `MultiStore` and related interfaces.
- `versiondb/tsrocksdb`, the implementation of `VersionDB` interface using the rocksdb user-defined timestamp.
- `versiondb/client`, implements the group of commands: `cronosd changeset` mainly for managing the legacy state migration process.
- `versiondb/extsort`, implements an external sorting algorithm, used in `convert-to-sst` command to sort the data before writing out sst files.
- `versiondb/memiavl`, an in-memory iavl implementation, currently used in `verify` command to replay and verify change set files, in the future, it'll evolves into a [full IAVL implementation](https://github.com/crypto-org-chain/cronos/blob/45162664a0f7bf25f4a3c14db01baf45d737a98a/versiondb/memiavl/README.md).

yihuang commented:
## `dump` performance

From genesis to `7130689`, 8cores, ssd
```
$ time --verbose cronosd changeset dump --home /chain/.cronosd/ --output data distribution acc authz bank capability cronos evidence evm feegrant feeibc feemarket gov ibc mint params slashing staking transfer upgrade
Percent of CPU this job got: 251%
Elapsed (wall clock) time (h:mm:ss or m:ss): 10:29:45
Maximum resident set size (kbytes): 19423828
$ du -sh data
3.8G	data/
```

## `convert-to-sst `

Convert the plain changeset format to rocksdb sst files with user-defined timestamp feature, spend 19m on testnet node (8cores), the result size: 5.4G, the generated sst files don't overlap with each other (don't trigger further compaction after ingested into versiondb).

The script is like this:
```python
import os
import sys
from datetime import datetime
from pathlib import Path
import subprocess
from concurrent.futures import ThreadPoolExecutor, as_completed

CRONOSD = './result/bin/cronosd'
DEFAULT_STORES = "distribution acc authz bank capability cronos evidence evm feegrant feeibc feemarket gov ibc mint params slashing staking transfer upgrade".split()

SOURCE = Path("./data")
DEST = Path("./sst")

# translate to around 2-3G peak ram usage for each task,
# tune it according to available ram and concurrency level
SORT_CHUNK_SIZE = 256*1024*1024


def run_store(executor, store, srcd, dstd):
    sst = dstd / f'{store}.sst'
    inputs = [src for src in (srcd / store).glob('block-*')]
    return executor.submit(subprocess.run, [CRONOSD, 'changeset', 'convert-to-sst', '--store', store, '--sorter-chunk-size', str(SORT_CHUNK_SIZE), sst] + inputs, check=True)

def main(stores, concurrency):
    DEST.mkdir(exist_ok=True, parents=True)
    with ThreadPoolExecutor(max_workers=concurrency) as executor:
        futs = []
        for store in stores:
            futs.append(run_store(executor, store, SOURCE, DEST))
        for fut in as_completed(futs):
            fut.result()

if __name__ == '__main__':
    try:
        stores = sys.argv[1].split()
    except IndexError:
        stores = DEFAULT_STORES
    main(stores, os.cpu_count())
```

## `ingest-sst`

Ingest the sst files into a rocksdb instance and set latest version, takes less than one second:

```
$ time cronosd changeset ingest-sst /chain/.cronosd/data/versiondb sst/*.sst --move-files --ingest-behind --maximum-version 6698242
```

JayT106 commented on go.mod:
```diff
@@ -192,20 +195,22 @@ replace (
-
-	// https://github.com/linxGnu/grocksdb/pull/100
-	github.com/linxGnu/grocksdb => github.com/yihuang/grocksdb v1.7.11-0.20230112025116-f41883a71db3
+	github.com/linxGnu/grocksdb => github.com/yihuang/grocksdb v1.7.11-0.20230120153422-b2f936d8375d
```
didn't see you submit the HyperClockCache changes to the upstream repo, will you submit it or keep it in `crypto-org-chain` repo?

yihuang replied:
Because it needs a pr in rocksdb itself.

yihuang replied:
https://github.com/facebook/rocksdb/pull/11110
I'll submit grocksdb PR after rocksdb PR merged.

yihuang replied:
don't need the fork anymore, reverted ClockCache for now.

JayT106 commented on versiondb/client/changeset.go:
```diff
@@ -0,0 +1,322 @@
+	binary.LittleEndian.PutUint16(buf[1:], uint16(len(pair.Key)))
+	copy(buf[3:], pair.Key)
+	if pair.Delete {
+		buf[0] = 1
```
maybe do return and then remove the else condition?

yihuang replied:
done

JayT106 commented on versiondb/client/changeset.go:
```diff
@@ -0,0 +1,322 @@
+}
+
+// IterateChangeSets iterate the change set files,
+func IterateChangeSets(
```
this function may take a long time to process, right? Should we add `context` if the user wants to interrupt the process?

JayT106 commented on versiondb/client/changeset.go:
```diff
@@ -0,0 +1,322 @@
+		err                error
+	)
+
+	for cont {
```
this for loop may take a long time to process, right? Should we add context if the user wants to interrupt the process?

yihuang replied:
it's currently only used in cli, user can Ctl+C directly.

JayT106 replied:
sure.

JayT106 commented on versiondb/client/changeset.go:
```diff
@@ -0,0 +1,322 @@
+		err                error
+	)
+
+	for cont {
```
the same question here.

yihuang replied:
it's currently only used in cli, user can Ctl+C directly, I think for enough for now?

JayT106 commented on versiondb/client/changeset.go:
```diff
@@ -0,0 +1,322 @@
+		if err == io.EOF {
+			// skipping empty files
+			continue
+		}
```
should handle other errors like failing to open the file?

yihuang replied:
good catch, fixed.

tomtau commented on versiondb/client/changeset.go:
```diff
@@ -0,0 +1,326 @@
+		return nil, fmt.Errorf("value length %d is too big", len(pair.Value))
+	}
+	offset := 3 + len(pair.Key)
+	binary.LittleEndian.PutUint32(buf[offset:], uint32(len(pair.Value)))
```
those uint16/uint32 casts should be safe because of those above checks, I assume

yihuang replied:
value length int32 should be safe, I also try to push for the `2^16` key length limit in sdk(https://github.com/cosmos/cosmos-sdk/pull/14645), but it seems cosmwasm has a bigger limit.
Maybe we should just use int32 for key as well.

yihuang replied:
I'll update change set format to use varint encoding, and stop worry about length limit here.

yihuang replied:
done, varint-uint64 encoding now.

tomtau commented on versiondb/extsort/sort.go:
```diff
@@ -0,0 +1,230 @@
+package extsort
```
is there a reason to have a custom external sort package instead of importing one?

yihuang replied:
It's simpler and cleaner than alternatives like https://github.com/lanrat/extsort, and no need to worry about third-party maintenence.
At the beginning I also tried to implement [tornament tree](https://en.wikipedia.org/wiki/K-way_merge_algorithm#Tournament_Tree) as recommended by wikipedia, but give up and just stick with heap in stdlib.

tomtau replied:
ok, maybe good to document this in the package main's file

yihuang replied:
done

tomtau reviewed (commented):
I had a quick look, but didn't review in detail (e.g. I skipped memiavl) -- are there any particular parts you'd like a more detailed review of?

yihuang commented on versiondb/client/changeset.go:
```diff
@@ -0,0 +1,326 @@
+	}
+
+	var versionHeader [16]byte
+	binary.LittleEndian.PutUint64(versionHeader[:], uint64(version))
+	binary.LittleEndian.PutUint64(versionHeader[8:], uint64(size))
```
> I assume there are no negative versions?

I think so, even `0` is reserved as special version for genesis in many places.

yihuang commented on versiondb/README.md:
```diff
@@ -0,0 +1,141 @@
+# VersionDB
```
Not yet, but some people are interested in our production experience with it.

yihuang commented on go.mod:
```diff
@@ -192,20 +195,22 @@ replace (
 	github.com/cosmos/cosmos-sdk => github.com/cosmos/cosmos-sdk v0.46.7
+
+	// https://github.com/yihuang/iavl/tree/optim_diff_backport
+	github.com/cosmos/iavl => github.com/yihuang/iavl v0.16.1-0.20230111025924-ae78cc50ad97
```
yes

yihuang replied:
backported in upstream, updated dependency.

yihuang commented on .github/workflows/build.yml:
```diff
@@ -15,6 +15,11 @@ jobs:
       - uses: actions/checkout@v3
+      - uses: cachix/install-nix-action@v18
+        with:
+          nix_path: nixpkgs=channel:nixos-22.11
```
This is needed to install rocksdb library to build `cronosd`, now librocksdb is always required to build `cronosd`.

yihuang replied:
it's not needed for default `make build` now, but we'll still try both (with or without rocksdb) in CI.

yihuang commented:
> I had a quick look, but didn't review in detail (e.g. I skipped memiavl) -- are there any particular parts you'd like a more detailed review of?

I think correctness-wise, it should be alright, several modules cross-check with each other, for example dump and verify, the versiondb content need to check with integration test, it seems pass all current integration tests fine.
Will need qa team to do more check after devops team did some node migrations.

calvinaco commented on versiondb/client/dump.go:
```diff
@@ -0,0 +1,266 @@
+	taskGroup    *pond.TaskGroupWithContext
+}
+
+// collect wait for the tasks to complete and concatinate the files into a single output file.
```
```suggestion
// collect wait for the tasks to complete and concatenate the files into a single output file.
```

yihuang commented on versiondb/client/changeset.go:
```diff
@@ -0,0 +1,326 @@
+func encodedSizeOfKVPair(pair iavl.KVPair) int {
+	size := 3 + len(pair.Key)
+	if !pair.Delete {
+		size += 4 + len(pair.Value)
```
The updated version have much less magic numbers now.

JayT106 commented on versiondb/client/dump.go:
```diff
@@ -0,0 +1,265 @@
+				return err
+			}
+
+			db, err := openReadOnlyDB(ctx.Viper.GetString(flags.FlagHome), server.GetAppDBBackend(ctx.Viper))
```
```suggestion
			db, err := openReadOnlyDB(ctx.Viper.GetString(flags.FlagHome), server.GetAppDBBackend(ctx.Viper))
			defer db.Close()
```

JayT106 commented on versiondb/client/dump.go:
```diff
@@ -0,0 +1,265 @@
+	}
+	defer fp.Close()
+
+	writer := snappy.NewBufferedWriter(fp)
```
maybe call defer `writer.Close()` here and then return `nil` at the end?

JayT106 replied:
But in line 186, it might just write the changeset into the snappy writer's buffer?

yihuang replied:
yeah, I mean if error happens, it doesn't matter too much if we flush the buffer or not.
And with defer we also better to handle the error `Flush()` returns, it'll be more than a simple line of code, right?

JayT106 replied:
yes, we will need to handle the error that happens in `defer`. So in line 186, looks like the writer didn't flush and return directly if I understand correctly, it needs to be flushed before return, right?

yihuang replied:
you mean the callback? it don't need to flush actively, we only need to flush the pending buffer at the very end.

JayT106 replied:
Ah, right, it's a callback.

JayT106 commented on versiondb/extsort/sort.go:
```diff
@@ -0,0 +1,235 @@
+	var err error
+	for _, chunkFile := range s.chunkFiles {
+		err1 := chunkFile.Close()
+		if err != nil {
```
should we append the error instead of assigning it if there is more than one close error?

yihuang replied:
not sure if it worth the trouble, I'll try to find a library to group errors if there's any.

yihuang replied:
done, `hashicorp/go-multierror` is good at this

JayT106 commented on versiondb/extsort/sort.go:
```diff
@@ -0,0 +1,235 @@
+
+// diffOffset returns the index of first byte that's different in two bytes slice,
+// the index is encoded as uint8, so the max supported length is `math.MaxUint8`,
+// that should be more than enough in our use case.
```
if we assume max supported length is `math.MaxUint8`. maybe do panic if we see the length exceeds the max?

yihuang replied:
out of date comment, it use varint encoding, so support int64 technically.

JayT106 commented on versiondb/memiavl/persisted_node.go:
```diff
@@ -0,0 +1,139 @@
+	if err2 != nil {
+		return err2
+	}
+	return err3
```
trying to close 3 files but here might return err1/err2 but other errors happen too.

yihuang replied:
done, `hashicorp/go-multierror` is good at this

JayT106 commented on versiondb/memiavl/persisted_node.go:
```diff
@@ -0,0 +1,139 @@
+// Value result is not defined for non-leaf node.
+func (node PersistedNode) Value() []byte {
+	valueOffset := binary.LittleEndian.Uint64(node.blobs.nodes[node.offset+OffsetValue:])
+	valueLen := uint64(binary.LittleEndian.Uint16(node.blobs.values[valueOffset : valueOffset+2]))
```
doesn't need `+2` like other fields?

yihuang replied:
fixed

JayT106 commented on versiondb/memiavl/persisted_node.go:
```diff
@@ -0,0 +1,139 @@
+
+func (node PersistedNode) Hash() []byte {
+	offset := node.offset + OffsetHash
+	return node.blobs.nodes[offset : offset+32]
```
```suggestion
	return node.blobs.nodes[offset : offset+ SizeHash]
```

JayT106 commented on versiondb/memiavl/snapshot.go:
```diff
@@ -0,0 +1,265 @@
+)
+
+func WriteSnapshot(root Node, version int64, nodesFile, keysFile, valuesFile, metadataFile string) error {
+	fpNodes, err := os.OpenFile(nodesFile, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o600)
```
make `os.O_WRONLY|os.O_CREATE|os.O_TRUNC` and `0o600` as const?

yihuang replied:
added a `createFile` utility to reduce repetition.

JayT106 commented on versiondb/memiavl/snapshot.go:
```diff
@@ -0,0 +1,265 @@
+	if len(nodes) == 0 {
+		// empty tree
+		if rootIndex != 0 {
+			return nil, 0, 0, errors.New("corrupted snapshot, nodes are empty but rootIndex is not zero")
```
fpNodes.Close()

JayT106 replied:
```suggestion
                         fpNodes.Close()
			return nil, 0, 0, errors.New("corrupted snapshot, nodes are empty but rootIndex is not zero")
```

JayT106 commented on versiondb/memiavl/snapshot.go:
```diff
@@ -0,0 +1,265 @@
+		valuesHandle: valuesHandle,
+	}
+
+	return blobs, version, rootIndex, nil
```
any concerns for not using `defer fpNodes.Close()`,  `defer fpKeys.Close()`, `defer fpValues.Close()`?

JayT106 commented on versiondb/tsrocksdb/comparator.go:
```diff
@@ -0,0 +1,52 @@
+
+// compare compares two internal keys with timestamp surfix, larger timestamp comes first.
+func compare(a []byte, b []byte) int {
+	ret := compareWithoutTS(a, true, b, true)
```
do we have other use cases for compareWithoutTS? If no, we can simplify the input to only a & b?

JayT106 commented on versiondb/tsrocksdb/comparator.go:
```diff
@@ -0,0 +1,52 @@
+// CreateTSComparator should be compatible with builtin timestamp comparator.
+func CreateTSComparator() *grocksdb.Comparator {
+	return grocksdb.NewComparatorWithTimestamp(
+		"leveldb.BytewiseComparator.u64ts", TimestampSize, compare, compareTS, compareWithoutTS,
```
should we call different name instead it?

yihuang replied:
I deliberate to use the same name as rocksdb builtin comparator, so the result database can still work with builtin tools like `ldb`/`sst_dump`.

JayT106 commented on versiondb/tsrocksdb/opts.go:
```diff
@@ -0,0 +1,63 @@
+
+	// 1G block cache
+	bbto.SetBlockSize(32 * 1024)
+	bbto.SetBlockCache(grocksdb.NewClockCache(1<<30, 32*1024))
```
Looks like the new rocksdb feature only support after v7.9 , need to update the build requirement in Cronos doc.

yihuang replied:
I was trying to experiment with it, don't have much visible changes, the dump process is still mainly IO bound, I'll convert back to LRU cache first to make this PR simpler.

JayT106 commented on versiondb/memiavl/README.md:
````diff
@@ -0,0 +1,114 @@
+  ...
+  ```
+
+  Value size is encoded in `int32`, so maximum key length supported is `2**32`, around 4G.
````
should be `2**31 - 1` because not `uint32`?

yihuang replied:
changed to `uint32`.

JayT106 commented on versiondb/memiavl/README.md:
````diff
@@ -0,0 +1,114 @@
+  ...
+  ```
+
+  Key size is encoded in `int16`, so maximum key length supported is `65536`. 
````
should be `2**15 - 1` if it's `int16`?

JayT106 commented on versiondb/README.md:
````diff
@@ -0,0 +1,141 @@
+
+```
+$ cronosd changeset dump --home /chain/.cronosd/ --output data distribution acc authz bank capability cronos evidence evm feegrant feeibc feemarket gov ibc mint params slashing staking transfer upgrade
+```
````
for UX, maybe we can add `all` instead of putting these modules?

JayT106 replied:
The following comments relate to UX and can be improved later I think.

yihuang replied:
I think it's not easy to detect all modules, because we need to include all modules since genesis, if some module is removed in previous upgrade, we can't find it in latest commit info.

tomtau replied:
maybe just a best-effort of hardcoding it statically?

yihuang replied:
> maybe just a best-effort of hardcoding it statically?

that could be misleading to user, if the list out dated, or there maybe difference in different stages of testnet and mainnet, it could be tedious to maintain in the future.

yihuang replied:
I think it better maintained at the doc site, similar to how we document all the breaking upgrades and binary versions.

yihuang replied:
https://github.com/crypto-org-chain/cronos/issues/826
Opened a separate issue to improve cmd UX later.

JayT106 commented on versiondb/README.md:
````diff
@@ -0,0 +1,141 @@
+```
+$ cronosd changeset verify --save-snapshot /tmp/snapshot data/acc/block-0.zz data/acc/block-1000000.zz data/acc/block-2000000.zz
+$ cronosd changeset verify --load-snapshot /tmp/snapshot data/acc/block-3000000.zz data/acc/block-4000000.zz data/acc/block-5000000.zz
+```
````
does the op/user requires to verify the changeset in each module manually? meaning verify it just assign the folder path then it will verify all files in the folder for better UX

yihuang replied:
depends on if user trust the data provider.

yihuang replied:
https://github.com/crypto-org-chain/cronos/issues/826
Opened a separate issue to improve cmd UX later.

JayT106 commented on versiondb/README.md:
````diff
@@ -0,0 +1,141 @@
+$ cronosd changeset convert-to-sst --store distribution ./sst/distribution.sst data/distribution/*.zz
+```
+
+`convert-to-sst` will do the sst file writing for a single store, you can wrap it in a simple script to run multiple stores in parallel. Here's an example Python script:
````
maybe put it directly into `scripts` folder. Or just do it in `convert-to-sst` cli directly, for better UX. In any case, we only convert a few modules instead all?

yihuang replied:
normally user should convert all modules, but yes, technically it's possible to only convert a subset of modules.

yihuang replied:
https://github.com/crypto-org-chain/cronos/issues/826
Opened a separate issue to improve cmd UX.

JayT106 commented:
reviewed most of the files except the tests.

yihuang commented:
I removed data block hash index and "OptimizeFiltersForHits" for versiondb, because it don't make as much sense there as the iavl tree.
Also enabled parallel compression, which could have some help on compaction speed.

mmsqe commented on versiondb/client/convert_to_sst.go:
```diff
@@ -0,0 +1,223 @@
+	return aVersion > bVersion
+}
+
+func cloneAppend(bz []byte, tail []byte) (res []byte) {
```
maybe can [cloneAppend](https://github.com/crypto-org-chain/cronos/blob/4900fe0333c7b70c18d18a39fda997e402adb5bc/versiondb/tsrocksdb/store.go#L195) reuse to util, but not sure if need clone since no mutate in both places

yihuang replied:
`append` itself can mutate the first parameter if there's capacity.

yihuang commented on versiondb/extsort/delta_encoding.go:
```diff
@@ -0,0 +1,89 @@
+		return err
+	}
+
+	if _, err := w.Write(key[shared:]); err != nil {
```
yes, delta encoding runs sequentially.

tomtau commented on versiondb/client/wrapreader.go:
```diff
@@ -0,0 +1,51 @@
+	closer io.Closer
+}
+
+// WrapReader wraps reader and closer together to create an new io.ReadCloser.
```
```suggestion
// WrapReader wraps reader and closer together to create a new io.ReadCloser.
```

tomtau commented on versiondb/memiavl/README.md:
```diff
@@ -0,0 +1,119 @@
+
+- Change set files can be splited with certain block ranges for incremental backup and restoration.
+
+- Historical files can be compressed with zlib, because it don't need to support random access.
```
```suggestion
- Historical files can be compressed with zlib, because it doesn't need to support random access.
```

tomtau commented on versiondb/memiavl/README.md:
```diff
@@ -0,0 +1,119 @@
+
+  Some integers are using `uint32`, should be enough in forseeable future, but could be changed to `uint64` to be safer.
+
+  The implementation will read the mmap-ed content in a zero-copy way, don't use extra node cache, just rely on OS page cache.
```
```suggestion
  The implementation will read the mmap-ed content in a zero-copy way, won't use extra node cache, it will only rely on the OS page cache.
```

tomtau commented on versiondb/memiavl/README.md:
```diff
@@ -0,0 +1,119 @@
+
+### VersionDB
+
+[VersionDB](../README.md) is to support query and iterating historical versions of key-values pairs, currently implemented with rocksdb's experimental user-defined timestamp feature, support query and iterate key-value pairs by version, it's an alternative way to support grpc query service, and much more compact than IAVL tree, similar in size with the compressed change set files.
```
```suggestion
[VersionDB](../README.md) is to support query and iterating historical versions of key-values pairs, currently implemented with rocksdb's experimental user-defined timestamp feature, support query and iterate key-value pairs by version, it's an alternative way to support grpc query service, and much more compact than IAVL trees, similar in size with the compressed change set files.
```

yihuang commented on gomod2nix.toml:
```diff
@@ -105,8 +108,9 @@ schema = 3
+    version = "v0.16.1-0.20230126105431-1bafbada9532"
+    hash = "sha256-eITR5aJLSgSUiVs3mGvrlsydkQIUTyqKl4djXefqn4Y="
+    replaced = "github.com/yihuang/iavl"
   [mod."github.com/cosmos/ibc-go/v5"]
```
done

yihuang commented on versiondb/client/changeset.go:
```diff
@@ -0,0 +1,358 @@
+}
+
+// uvarintSize returns the size (in bytes) of `num` encoded as a unsigned varint.
+func uvarintSize(num uint64) int {
```
done, referenced `binary.PutUvarint` directly.

yihuang commented on versiondb/client/convert_to_sst.go:
```diff
@@ -0,0 +1,223 @@
+	version := item[keyLen : keyLen+8]
+	delete_ := item[keyLen+8] == 1
+	if !delete_ {
+		value = item[keyLen+9 : len(item)-SizeKeyLength]
```
yeah, refactored a little bit to reduce magic numbers.

yihuang commented:
it seems can't disable the `typecheck` linter in CI? it don't reproduce locally, will ignore for now.
related: https://github.com/golangci/golangci-lint-action/issues/624

yihuang commented:
> it seems can't disable the `typecheck` linter in CI? it don't reproduce locally, will ignore for now. related: [golangci/golangci-lint-action#624](https://github.com/golangci/golangci-lint-action/issues/624)

finally figured out, typecheck error just means it fails to compile somewhere, it's CGO dependencies not setup in our case.
